Support Windows

daipom opened the pull request:
This adds Windows support.

I added default configuration values for Windows, and separated the parsing logic.

The command for Windows had to be complicated, 
but I made it possible to change the keys without setting the command in Windows
so that users don't have to understand Windows command.

daipom committed:
Add Widows support

daipom committed:
Fix to always call IO close

daipom committed:
Fix to use Fluent module functions to identify OS type

daipom committed:
Just Refactoring

daipom committed:
Fix to format 'start_time' output in Windows

daipom committed:
Update README about Windows

daipom committed:
Add Windows test

daipom committed:
Refactor test

daipom committed:
Fix Windows default keys

- Windows keys now point directly to the properties of .NET object.

daipom committed:
Fix Windows command about datetime format to be simpler

daipom committed:
Add elapsed time to Windows like Linux and Mac

ashie commented:
@y-ken @cosmo0920 Could you kindly take a look this?

y-ken commented:
Looks like so great work however I cannot verify on windows, could anybody check on windows 10?

cosmo0920 commented:
I'll take a look on this.

daipom commented:
@y-ken @cosmo0920 Thank you very much!

y-ken commented on README.md:
```diff
@@ -90,6 +96,38 @@ $ tail -f /var/log/td-agent/td-agent.log
 
+### About Windows
+
+Default `command` in Windows below is complicated, but you can fix `keys` without fixing `command`, unless you specify a key other than the default.
```
I guess this is more user-friendly English
```suggestion
Default `command` preset for Windows provides many of keys as below. Generally, you can pick up the columns with `keys` option.
If you need additional keys, consider to update `command` option
```

y-ken commented:
If some windows user using Fluentd on Windows 8 , 7 or Vista or earlier version, How about to add a note as like "this is working after Windows *** or later"?
or, do we need to specify powershell version instead?

cosmo0920 committed:
Provide PowerShell 7 command choice

Signed-off-by: Hiroshi Hatake <hatake@calyptia.com>

cosmo0920 committed:
Add descriptions for PowerShell command name choices

Signed-off-by: Hiroshi Hatake <hatake@calyptia.com>

cosmo0920 commented:
> If some windows user using Fluentd on Windows 8 , 7 or Vista or earlier version, How about to add a note as like "this is working after Windows *** or later"?
> or, do we need to specify powershell version instead?

I think that we should describe which versions we used for confirming this feature on Windows.

e.g) Windows 10 w/ PowerShell 5.1 and Windows 10 w/ PowerShell 7.1 is confirmed to be working well.

cosmo0920 commented:
The reasons why I happened to find that this plugin w/ PowerShell 7 is not workable, I'm usually using PowerShell 7 is the default shell of Windows Terminal....

cosmo0920 reviewed (commented):
I added a few commits to support newer PowerShell environment such as PowerShell 7.
Otherwise, it looks good to me.
Thanks for your work! :punch:

cosmo0920 committed:
Use more user-friendly English

Co-authored-by: Y.Kentaro <y.ken.studio@gmail.com>

cosmo0920 committed:
Add confirmed version combinations

But this is not meant other version combinations are "not workable".
This is just our confirmation environment information.

Signed-off-by: Hiroshi Hatake <hatake@calyptia.com>

cosmo0920 commented:
I'd merged this PR into master.
@y-ken Could you release the new version of this plugin?
I think that the next version should be v0.2.0 or something.

cosmo0920 commented:
@ashie or @kenhys, Could you add using your confirmation environment information under https://github.com/y-ken/fluent-plugin-watch-process#about-windows?
I don't use Windows server platform to confirm this PR behavior. If you use such platform, please add their information there.

y-ken commented:
I'm preparing to release v0.2.0.
And... Good news, I have added you in a owner list
![image](https://user-images.githubusercontent.com/1734549/120099138-e490e400-c174-11eb-8717-7cb419b9c7a0.png)

y-ken commented:
released v0.2.0

https://rubygems.org/gems/fluent-plugin-watch-process

ashie commented:
> And... Good news, I have added you in a owner list

Oh! I've confirmed it. Thanks!

daipom commented:
@y-ken @cosmo0920 @ashie 
Thank you for the review and fix! I'm so happy this fix was merged!
I did not check enough about the Powershell version. Thank you very much for the fix.

cosmo0920 commented:
Thanks. I'd received the ownership of this plugin. :punch:
